lookup_estim.go

```diff
@@ -0,0 +1,302 @@
+package dht
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/libp2p/go-libp2p-kad-dht/netsize"
+	"github.com/libp2p/go-libp2p-kad-dht/qpeerset"
+	kb "github.com/libp2p/go-libp2p-kbucket"
+	"github.com/libp2p/go-libp2p/core/peer"
+	ks "github.com/whyrusleeping/go-keyspace"
+	"gonum.org/v1/gonum/mathext"
+)
+
+const (
+	// OptProvIndividualThresholdCertainty describes how sure we want to be that an individual peer that
+	// we find during walking the DHT actually belongs to the k-closest peers based on the current network size
+	// estimate.
+	OptProvIndividualThresholdCertainty = 0.9
+
+	// OptProvSetThresholdStrictness describes the probability that the set of closest peers is actually further
+	// away then the calculated set threshold. Put differently, what is the probability that we are too strict and
+	// don't terminate the process early because we can't find any closer peers.
+	OptProvSetThresholdStrictness = 0.1
+
+	// OptProvReturnRatio corresponds to how many ADD_PROVIDER RPCs must have completed (regardless of success)
+	// before we return to the user. The ratio of 0.75 equals 15 RPC as it is based on the Kademlia bucket size.
+	OptProvReturnRatio = 0.75
+)
+
+type addProviderRPCState int
+
+const (
+	Sent addProviderRPCState = iota + 1
+	Success
+	Failure
+)
+
+type estimatorState struct {
+	// context for all ADD_PROVIDER RPCs
+	putCtx context.Context
+
+	// reference to the DHT
+	dht *IpfsDHT
+
+	// the most recent network size estimation
+	networkSize float64
+
+	// a channel indicating when an ADD_PROVIDER RPC completed (successful or not)
+	doneChan chan struct{}
+
+	// tracks which peers we have stored the provider records with
+	peerStatesLk sync.RWMutex
+	peerStates   map[peer.ID]addProviderRPCState
```

Comment:
I didn't use a `sync.Map` because I'd lose type safety and casting values is annoying. Not sure, I'd pull in an extra dependency here - although you told me that you've audited that.

Reply:
Fair enough, let's not change this unless we have a profile showing contention.